auth/common.cc

```diff
@@ -113,4 +118,104 @@ ::service::query_state& internal_distributed_query_state() noexcept {
     return qs;
 }
 
+static future<> announce_mutations_with_guard(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    auto group0_cmd = group0_client.prepare_command(
+        ::service::write_mutations{
+            .mutations{std::move(muts)},
+        },
+        group0_guard,
+        "auth: modify internal data"
+    );
+    return group0_client.add_entry(std::move(group0_cmd), std::move(group0_guard), as);
+}
+
+static future<> announce_mutations_maybe_split(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        std::vector<canonical_mutation> muts,
+        ::service::group0_guard group0_guard,
+        seastar::abort_source* as) {
+    auto begin = muts.begin();
+    auto end = muts.end();
+    size_t batch_size  = muts.size();
+    auto remaining = batch_size;
+    while (remaining) {
+        try {
+            co_await announce_mutations_with_guard(qp, group0_client, {begin, end}, std::move(group0_guard), as);
+            remaining -= (end - begin);
+            begin = end;
+        } catch (raft::command_is_too_big_error&) {
+            batch_size /= 2;
+            if (batch_size == 0) {
+                std::throw_with_nested(std::runtime_error("raft add entry failed with batch size 1"));
+            }
+            // when retrying we don't change begin and end gets truncated below
+        }
+        if (remaining) {
+            end = begin + std::min(batch_size, remaining);
+            // FIXME: we need a way to reuse single guard when retry
+            // doesn't require command data regeneration as in this case
+            // otherwise we may violate operation consistency guarantee
+            group0_guard = co_await group0_client.start_operation(as);
+        }
+    }
+}
+
+future<> announce_mutations_with_batching(
+        cql3::query_processor& qp,
+        ::service::raft_group0_client& group0_client,
+        global_schema_ptr global_schema,
+        ::service::group0_guard group0_guard,
+        std::function<mutations_generator()> gen,
+        seastar::abort_source* as) {
+    // account for command's overhead, it's better to use smaller threshold than constantly bounce off the limit
+    const size_t memory_threshold = group0_client.max_command_size() * 0.75;
+    size_t memory_usage = 0;
+    auto schema = global_schema.get();
+    std::vector<canonical_mutation> muts;
+
+    auto g = gen();
+    while (auto mut = co_await g()) {
+        muts.push_back(canonical_mutation{*mut});
+        memory_usage += muts.back().representation().size();
+        if (memory_usage >= memory_threshold) {
+            if (!group0_guard) {
```

Comment:
This is relying on internal implementation detail of guard (move operator implementation), it's basically use-after-move. There is no guarantee given by the interface of guard that `std::move`ing from it leaves a disengaged guard. I see no such comment at least.

BTW. I think it's a mistake to make guard basically an optional object... Now whenever we need to write code using guards, we need to reason about it as if it was optional.


Anyway, what's the purpose of the guard in this function, if we're dropping it between each mutation? There is no protection from concurrent modifications. This is equivalent to sending mutations unguarded.

We're also generating the mutations outside guard. What timestamps are these mutations using if not those from a guard?

I don't get it.

Reply:
see discussion below https://github.com/scylladb/scylladb/pull/16578#discussion_r1492720411